### PR TITLE
Memory leak fixes, again!

### PR DESCRIPTION
**Pull request by MaxKellermann**
Another big bunch of memory leak fixes, and several wrong/missing references which could make Zeek crash. These crashes never occurred because Zeek never happens to actually free those objects, due to the numerous memory leaks which keep those objects referenced. The more memory leaks we fix, the more likely will those missing references cause real crashes...

### CODE REVIEW

File: src/option.bif [hunk 1]
```diff
@@ -205,6 +205,8 @@ function Option::set_change_handler%(ID: string, on_change: any, priority: int &
 		return val_mgr->GetBool(0);
 		}
 
-	i->AddOptionHandler(on_change->Ref()->AsFunc(), -priority);
+	auto *func = on_change->AsFunc();
```
Comment (jsiwek) [inline src/option.bif]:
```suggestion
	auto func = on_change->AsFunc();
```
style nit

File: src/option.bif [hunk 2]
```diff
@@ -18,7 +18,7 @@ static bool call_option_handlers_and_set_value(StringVal* name, ID* i, Val* val,
 			bool add_loc = handler_function->FType()->AsFuncType()->ArgTypes()->Types()->length() == 3;
 			val_list vl(2 + add_loc);
 			vl.push_back(name->Ref());
-			vl.push_back(val);
+			vl.push_back(val->Ref());
```
Comment (jsiwek) [inline src/option.bif]: This change doesn't look right and likely explains at least some CI memcheck failures reporting new leaks.  The function call does consume a reference, but it also returns a new `Val` directly into `val` on return that we take ownership of -- either the next iteration consumes that or the `Unref(val)` at the end of the function will if no more iterations.

File: src/analyzer/protocol/login/Login.cc [hunk 3]
```diff
@@ -564,6 +564,7 @@ void Login_Analyzer::AddUserText(const char* line)
 		if ( ++user_text_last == MAX_USER_TEXT )
 			user_text_last = 0;
 
+		delete[] user_text[user_text_last];
```
Comment (jsiwek) [inline src/analyzer/protocol/login/Login.cc]: We came to a conclusion a bit ago that this wasn't right: https://github.com/zeek/zeek/pull/756#discussion_r372101289

Maybe time to leave a comment so we don't accidentally revisit again later?

File: src/parse.y [hunk 4]
```diff
@@ -1684,15 +1696,21 @@ for_head:
 					key_var->Error("global variable used in for loop");
 				}
 			else
+				{
+				Unref(key_var);
```
Comment (jsiwek) [inline src/parse.y]: `key_var` is null on this branch so this unref is superfluous?

File: src/parse.y [hunk 5]
```diff
@@ -1684,15 +1696,21 @@ for_head:
...
+				}
 
 			if ( val_var )
 				{
 				if ( val_var->IsGlobal() )
 					val_var->Error("global variable used in for loop");
 				}
 			else
+				{
+				Unref(val_var);
```
Comment (jsiwek) [inline src/parse.y]: `val_var` is null on this branch so unref is superfluous?

File: src/parse.y [hunk 6]
```diff
@@ -1714,7 +1732,10 @@ for_head:
...
```
Comment (jsiwek) [inline src/parse.y]: `val_var` already null, so superfluous unref again?

File: src/parse.y [hunk 7]
```diff
@@ -1745,6 +1766,7 @@ local_id:
 
 			else
 				{
+				Unref($$);
```
Comment (jsiwek) [inline src/parse.y]: Already null so no need to unref?

File: src/parse.y [hunk 8]
```diff
@@ -1791,6 +1813,7 @@ global_or_event_id:
...
```
Comment (jsiwek) [inline src/parse.y]: Already null, no need to unref?

File: src/analyzer/protocol/login/Login.cc (hunk 3 above)
Comment (MaxKellermann) [reply src/analyzer/protocol/login/Login.cc]: Oops, this was still on my stgit stack, and when I assembled this PR yesterday, it looked so correct... sorry.

File: src/option.bif (hunk 2 above)
Comment (MaxKellermann) [reply src/option.bif]: You're right, the code is indeed correct after all! This time, my confusion was just my confusion and not a bug...

File: src/parse.y (hunk 4 above)
Comment (MaxKellermann) [reply src/parse.y]: Thanks, 4 unnecessary `Unref(nullptr)` removed. 6 memory leak fixes remain in the commit.

**Comment by MaxKellermann** [comment]
Dropped 2 wrong commits and 4 non-fixes (`Unref()` on variables which were known to be `nullptr`).

File: src/option.bif (hunk 1 above)
Comment (MaxKellermann) [reply src/option.bif]: Fixed

**Comment by MaxKellermann** [comment]
memcheck failure - I'll check and fix that later.

**Comment by jsiwek** [comment]
> memcheck failure - I'll check and fix that later.

Likely can remove 2d52e319eacadd8f897e963bdcd324a1f82d099f to fix it.  Think the `Expr*` passed in should be treated as owned by the "global AST" and if anyone in `make_var` ends up storing it for later use, they should `Ref()`.

**Comment by MaxKellermann** [comment]
> if anyone in `make_var` ends up storing it for later use, they should `Ref()`.

Which means the function is currently bugged in any case, because:
https://github.com/zeek/zeek/blob/48fba11c5158285926a223f539ac2155d0407817/src/Var.cc#L171
.. takes ownership of the caller's reference to `init`, and will free that reference (but the new pointer created here will never be freed or "owned" by anybody).
That line is what made me believe that the function adopts ownership. As usual, Zeek code is very inconsistent, and deducing reference ownerships from existing code is often ambiguous, because existing code is wrong too often.

**Comment by MaxKellermann** [comment]
I found the problem with https://github.com/zeek/zeek/commit/2d52e319eacadd8f897e963bdcd324a1f82d099f - my commit is not buggy, after all - but `add_and_assign_local()` is buggy:
https://github.com/zeek/zeek/blob/48fba11c5158285926a223f539ac2155d0407817/src/Var.cc#L263-L268
This function passes the ownership for `init` to `make_var()` and keeps on using `init`, passing its ownership again to `AssignExpr::AssignExpr()`.

There are two code paths in `make_var()` which imply that the function adopts reference ownership:
https://github.com/zeek/zeek/blob/48fba11c5158285926a223f539ac2155d0407817/src/Var.cc#L171
and
https://github.com/zeek/zeek/blob/48fba11c5158285926a223f539ac2155d0407817/src/Var.cc#L160

I tried rewriting my commit to your suggestion (take another reference if the function needs it, instead of adopting ownership), but that way, the function gets more complicated due to this line:
https://github.com/zeek/zeek/blob/48fba11c5158285926a223f539ac2155d0407817/src/Var.cc#L171
This line needs a reference to the `input` parameter, and sets `init` to a newly created object. Now if this variable holds a pointer to a reference owned by somebody else, who will then take care for releasing this reference only we currently own? If we unref it right now, it will be destroyed immediately. This could be worked around with another variable, but then it just gets more complicated.

All of this complexity will be gone after the migration to `IntrusivePtr<>`, so let's keep it as simple as possible until then...

**Pull request merged**